src/rotest/api/common/models.py

```diff
@@ -0,0 +1,186 @@
+            resource to be locked.
+    """
+    PROPERTIES = [
+        ArrayField(name="descriptors", items_type=ResourceDescriptorModel,
+                   required=True)
+    ]
+
+
+class ReleaseResourcesParamsModel(AbstractAPIModel):
+    """Release the given resources names.
+
+    Args:
+        resources (list): list of str. the names of the resources to be
+            released.
+    """
+    PROPERTIES = [
+        ArrayField(name="resources", items_type=StringField("resource_name"),
+                   example=["calc1", "calc2"], required=True)
+    ]
+
+
+class TestResultModel(AbstractAPIModel):
+    """Describes the result of a test.
+
+    Args:
+        result_code (number): the end result code of the test.
+        info (str): additional info of th test run.
+    """
+    PROPERTIES = [
+        NumberField(name="result_code", required=True),
+        StringField(name="info", required=True)
+    ]
+
+
+class TestControlOperationParamsModel(AbstractAPIModel):
+    """Generic data structure of test control operations.
+
+    Args:
+        token (str): the session token of the current test run.
+        test_id (number): the relevant test to be influenced by the operation.
+    """
+    PROPERTIES = [
+        StringField(name="token", required=True),
+        NumberField(name="test_id", required=True)
+    ]
+
+
+class UpdateResourcesParamsModel(AbstractAPIModel):
+    """Update the given resources test data.
+
+    Args:
+        descriptors (list): list of ResourceDescriptorModel. the resources
+            to be updated.
+        test_details (TestControlOperationParamsModel): the details of the
+            relevant session datas.
+    """
+    PROPERTIES = [
+        ArrayField(name="descriptors", items_type=ResourceDescriptorModel,
+                   required=True),
+        ModelField(name="test_details",
+                   model=TestControlOperationParamsModel, required=True)
+    ]
+
+
+class AddTestResultParamsModel(AbstractAPIModel):
+    """Add a test result to a given test.
+
+    Args:
+        test_details (TestControlOperationParamsModel): the details of the
+            relevant session datas.
+        result (TestResultModel): the result of the test.
+    """
+    PROPERTIES = [
+        ModelField(name="test_details",
+                   model=TestControlOperationParamsModel, required=True),
+        ModelField(name="result", model=TestResultModel, required=True)
+    ]
+
+
+class TestModel(AbstractAPIModel):
+    """Test model structure.
+
+    Args:
+        id (number): the id of the test.
+        name (str): the name of the test.
+        class (str): the classname of the test.
+        subtests (list): list of TestModel. the subtests of the current test.
+    """
+    TITLE = "Test"
+    PROPERTIES = [
+        NumberField(name="id"),
+        StringField(name="name"),
+        StringField(name="class"),
+        ArrayField(name="subtests",
+                   items_type=DynamicType("TestModel",
```

Comment:
It is impassible in the launguge to use a referance of something before declaring it.
Here the class is on the declaration state, therefore it cannot find itself - because it is yet to be declared.
To solve that, there are 2 ways to approch it -
 1- to create a new type which dynamically imports the model
 2- using eval("TestModel")

I think 1 is a better option - as a closure.